File: bindings/dotnet/Xui/NavigationAnimation.cs
using System.Runtime.InteropServices;

namespace Xui;

public sealed partial class NavigationView
{
    public uint Duration
    {
        get { Window.Guard(); Window.Check(Native.NavigationDurationGet(Handle, out uint value)); return value; }
        set { Window.Guard(); Window.Check(Native.NavigationDurationSet(Handle, value)); }
    }
    public bool Animating
    {
        get { Window.Guard(); Window.Check(Native.NavigationAnimating(Handle, out uint value)); return value != 0; }
    }
    public NavigationView SetDuration(uint milliseconds) { Duration = milliseconds; return this; }
}

internal static partial class Native
{
    [LibraryImport("xui", EntryPoint = "xui_navigation_view_set_duration")]
    internal static partial int NavigationDurationSet(ulong target, uint milliseconds);
    [LibraryImport("xui", EntryPoint = "xui_navigation_view_get_duration")]
    internal static partial int NavigationDurationGet(ulong target, out uint milliseconds);
    [LibraryImport("xui", EntryPoint = "xui_navigation_view_get_animating")]
    internal static partial int NavigationAnimating(ulong target, out uint animating);
}

File: bindings/rust/xui/src/navigation_animation.rs
use super::*;

impl NavigationView {
    pub fn set_duration(&self, milliseconds: u32) -> Result<()> {
        check(unsafe { sys::xui_navigation_view_set_duration(self.handle, milliseconds) })
    }
    pub fn duration(&self) -> Result<u32> {
        let mut value = 0;
        check(unsafe { sys::xui_navigation_view_get_duration(self.handle, &mut value) })?;
        Ok(value)
    }
    pub fn animating(&self) -> Result<bool> {
        let mut value = 0;
        check(unsafe { sys::xui_navigation_view_get_animating(self.handle, &mut value) })?;
        Ok(value != 0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn navigation_animation_contract() -> Result<()> {
        let window = Window::new("Navigation duration", 600., 400.)?;
        let view = window.navigation_view("Navigation")?;
        assert_eq!(view.duration()?, 0);
        assert!(!view.animating()?);
        for duration in [0, 180, 10000] {
            view.set_duration(duration)?;
            assert_eq!(view.duration()?, duration);
        }
        for duration in [10001, u32::MAX] {
            assert!(view.set_duration(duration).is_err());
            assert_eq!(view.duration()?, 10000);
        }
        assert_ne!(
            unsafe { sys::xui_navigation_view_get_duration(view.handle, std::ptr::null_mut()) },
            0
        );
        let input = window.text_input("Not navigation")?;
        let mut value = 0;
        assert_ne!(
            unsafe { sys::xui_navigation_view_get_animating(input.handle, &mut value) },
            0
        );
        Ok(())
    }
}
